libraries/AP_AIS/AP_AIS.cpp

```diff
@@ -0,0 +1,765 @@
+    // @Units: s
+    // @Range: 1 2000
+    // @User: Advanced
+    AP_GROUPINFO("TIME_OUT", 3, AP_AIS, _time_out, TIME_OUT_DEFAULT),
+
+    // @Param: OPTIONS
+    // @DisplayName: AIS options
+    // @Description: Bitmask of AIS options
+    // @Bitmask: 0:Log all AIVDM messages,1:Log only unsupported AIVDM messages,2:Log decoded messages
+    // @User: Advanced
+    AP_GROUPINFO("OPTIONS", 4, AP_AIS, _options, AIS_OPTIONS_LOG_UNSUPPORTED_RAW),
+
+    AP_GROUPEND
+};
+
+// constructor
+AP_AIS::AP_AIS() :
+    _list(8)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+// return true if AIS is enabled
+bool AP_AIS::enabled() const
+{
+    return _type != AIS_NONE;
+}
+
+// Initialize the AIS object and prepare it for use
+void AP_AIS::init()
+{
+    if (!enabled()) {
+        return;
+    }
+
+    _uart = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_AIS, 0);
+    if (_uart == nullptr) {
+        return;
+    }
+
+    _uart->begin(AP::serialmanager().find_baudrate(AP_SerialManager::SerialProtocol_AIS, 0));
+}
+
+// update AIS, expected to be called at 20hz
+void AP_AIS::update()
+{
+    if (!_uart || !enabled()) {
+        return;
+    }
+
+    // read any available lines
+    int16_t nbytes = _uart->available();
+    while (nbytes-- > 0) {
+        char c = _uart->read();
+        if (decode(c)) {
+            const bool log_all = (_options & AIS_OPTIONS_LOG_ALL_RAW) != 0;
+            const bool log_unsupported = ((_options & AIS_OPTIONS_LOG_UNSUPPORTED_RAW) != 0) && !log_all; // only log unsupported if not logging all
+
+            if (_incoming.total  > AIVDM_BUFFER_SIZE)  {
+                // no point in trying to decode it wont fit
+                if (log_all || log_unsupported) {
+                    log_raw(&_incoming);
+                }
+                break;
+            } else if (log_all) {
+                log_raw(&_incoming);
+            }
+
+            if (_incoming.num == 1 && _incoming.total == 1) {
+                // single part message
+                if (!payload_decode(_incoming.payload) && log_unsupported) {
+                    // could not decode so log
+                    log_raw(&_incoming);
+                }
+            } else if (_incoming.num == _incoming.total) {
+                // last part of a multi part message
+                uint8_t index = 0;
+                uint8_t msg_parts[_incoming.total - 1];
+                for  (uint8_t i = 0; i < AIVDM_BUFFER_SIZE; i++) {
+                    // look for the rest of the message from the start of the buffer
+                    // we assume the mesage has be received in the correct order
+                    if (_AIVDM_buffer[i].num == (index + 1) && _AIVDM_buffer[i].total == _incoming.total && _AIVDM_buffer[i].ID == _incoming.ID) {
+                        msg_parts[index] = i;
+                        index++;
+                        if (index == _incoming.num) {
```

Comment:
```suggestion
                        if (index >= _incoming.num) {
```

It's a can't-happen, but it will limit memory overwrites in case of a bug being introduced later

Reply:
done